infrastructures/kubernetes/src/main/java/org/eclipse/che/workspace/infrastructure/kubernetes/cache/jpa/JpaKubernetesRuntimeStateCache.java

```diff
@@ -186,4 +225,31 @@ protected void doPutIfAbsent(KubernetesRuntimeState runtimeState) {
     em.persist(runtimeState);
     em.flush();
   }
+
+  @Singleton
+  public static class RemoveKubernetesRuntimeBeforeWorkspaceRemoved
+      extends CascadeEventSubscriber<BeforeWorkspaceRemovedEvent> {
+
+    @Inject private EventService eventService;
+    @Inject private JpaKubernetesRuntimeStateCache k8sRuntimes;
+
+    @PostConstruct
+    public void subscribe() {
+      eventService.subscribe(this, BeforeWorkspaceRemovedEvent.class);
+    }
+
+    @Override
+    public void onCascadeEvent(BeforeWorkspaceRemovedEvent event) throws Exception {
+      Optional<KubernetesRuntimeState> k8sRuntimeStateOpt =
+          k8sRuntimes.find(event.getWorkspace().getId());
+      if (k8sRuntimeStateOpt.isPresent()) {
+        // It is not normal case when non STOPPED workspace is removed.
+        // Need to log error to investigate why it may happen
+        // and clean up existing runtime not to lock removing of workspace
+        LOG.error("Workspace is removed but Kubernetes runtime state exists");
```

Comment:
Please add more details about this event. Maybe WS_ID, user ID, etc. Something that would help us to figure out why this happens.

Reply:
Thanks. Nice catch.